### ISSUE TITLE
early_stopping_rounds

### ISSUE THREAD

**Issue by kryptonite0** [stanfordmlgroup/ngboost#37]
Would it make sense to implement `early_stopping_rounds` like in LGBM or XGB? If so, I'm happy to contribute to the issue.

**Comment by avati** [comment]
We would certainly welcome such a contribution!

**Comment by kryptonite0** [comment]
Thank you. Just to check that I understand the code correctly, can you please confirm the following? At the moment, the model stops training if a validation set is provided, if at least 11 iterations have been completed, and if the average validation loss of the last 5 iterations (`[-5:]`) is bigger than the average validation loss of the previous 5 iterations (`[-10:-5]`), under the assumption that the loss is expected to decrease. In parallel, there is also the gradient tolerance stopping, but it's not related to this issue.

**Comment by avati** [comment]
That is correct. Could you briefly explain how xgboost's early_stopping_rounds differs? We would like to provide as similar interface to other libraries as possible.

**Comment by kryptonite0** [comment]
I believe it is implemented as a callback, and it's a bit more structured than NGBoost. It allows you to provide the model with more than one validation set. If there's more than one, it will use only the last. It also allows you to provide more than one validation metric, and - again - it will use only the last metric for early stopping. Depending on the validation metric to check against, it will try maximise/minimise it. If early stopping is enabled, the model records the best iteration observed during training.

**Comment by chaupmcs** [comment]
Just a small mention on LightGBM's early stopping.
LightGBM uses all the metrics for early_stopping as default. In addition, it has a parameter named  `first_metric_only` for using only one metric (first one) to calculate early stopping.

**Comment by kryptonite0** [comment]
Given that NGBoost doesn't feature multiple metrics or validation sets yet, I would just implement a simple `if` condition that, in case `early_stopping_rounds` are set to a positive integer, checks whether the loss has improved over the last `early_stopping_rounds` and terminates or continues the training accordingly. What do you think?

**Comment by alejandroschuler** [comment]
Closed via https://github.com/stanfordmlgroup/ngboost/commit/32e9736323b282f524274b7d88b05f0e84f987c8
